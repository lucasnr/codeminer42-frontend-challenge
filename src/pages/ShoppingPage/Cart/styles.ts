import styled from 'styled-components';

export const Container = styled.section`
	background-color: rgba(0, 0, 0, 0.075);
	border-radius: 0.5rem;
	overflow: hidden;
`;

export const Title = styled.h2`
	background-color: ${({ theme }) => theme.colors.secondary};
	color: ${({ theme }) => theme.text.secondary};
	font-size: 1.5rem;
	font-weight: 600;
	padding: 0.5rem;
	text-align: center;
`;

export const Content = styled.div`
	padding: 1rem;
`;

export const Discount = styled.form`
	display: flex;

	input {
		border: 0.0625rem solid rgba(0, 0, 0, 0.5);
		border-radius: 0.25rem;
		font-size: 0.875rem;
		font-weight: 400;
		margin-right: 1rem;
		padding: 0.625rem 1rem;
		width: 100%;
	}

	button {
		background-color: ${({ theme }) => theme.colors.primary};
		border: none;
		border-radius: 0.5rem;
		color: ${({ theme }) => theme.text.secondary};
		font-size: 0.875rem;
		font-weight: 500;
		padding: 0.625rem 1rem;
		text-transform: uppercase;
	}
`;

export const Totals = styled.ul`
	li {
		align-items: center;
		border-top: 0.0625rem solid rgba(0, 0, 0, 0.125);
		display: flex;
		font-size: 1rem;
		justify-content: space-between;
		padding: 1rem;

		&:last-child {
			font-weight: 700;
		}
	}
`;
