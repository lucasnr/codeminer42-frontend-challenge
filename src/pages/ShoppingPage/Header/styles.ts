import styled from 'styled-components';

export const Wrapper = styled.header`
	background-color: ${({ theme }) => theme.colors.secondary};
	color: ${({ theme }) => theme.text.secondary};
`;

export const Container = styled.div`
	align-items: center;
	display: flex;
	justify-content: space-between;
	margin: 0px auto;
	max-width: 62rem;
	padding: 1rem;
	width: 100%;
`;

export const Title = styled.h1`
	font-size: 2rem;
	font-weight: 700;
`;

export const User = styled.div`
	align-items: center;
	display: flex;
	font-size: 1.125rem;
	font-weight: 500;

	img {
		border: 0.125rem solid ${({ theme }) => theme.text.secondary};
		border-radius: 50%;
		margin-right: 1rem;
		height: 2.75rem;
		width: 2.75rem;
	}
`;
