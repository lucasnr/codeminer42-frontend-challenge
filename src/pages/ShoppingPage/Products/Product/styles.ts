import styled from 'styled-components';

export const Container = styled.div`
	border-radius: 0.5rem;
	overflow: hidden;

	img {
		background-color: rgba(0, 0, 0, 0.1);
		display: block;
		height: 9rem;
		object-fit: cover;
		object-position: center;
		width: 100%;

		@media (min-width: 576px) {
			height: 7.5rem;
		}
	}
`;

export const Info = styled.div`
	background-color: ${({ theme }) => theme.background.primary};
	padding: 1rem;

	> * {
		margin-bottom: 0.5rem;
	}
`;

export const Name = styled.h3`
	font-size: 1rem;
	font-weight: 600;
`;

export const Price = styled.h4`
	display: flex;
	font-size: 1rem;
	font-weight: 400;

	.dot {
		background-color: currentColor;
		border-radius: 50%;
		content: ' ';
		display: block;
		height: 0.25rem;
		margin: auto 1ch;
		width: 0.25rem;
	}
`;

export const Button = styled.button`
	background-color: ${({ theme }) => theme.colors.primary};
	border: none;
	color: ${({ theme }) => theme.text.secondary};
	font-size: 1rem;
	font-weight: 500;
	padding: 0.625rem 1rem;
	text-transform: uppercase;
	width: 100%;
`;
