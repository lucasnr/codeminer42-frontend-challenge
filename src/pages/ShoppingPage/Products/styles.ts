import styled, { keyframes } from 'styled-components';

export const Container = styled.section`
	display: grid;
	grid-gap: 2rem;
	grid-template-columns: repeat(auto-fit, minmax(12rem, 1fr));

	@media (min-width: 768px) {
		margin-bottom: auto !important;
	}
`;

export const Message = styled.span`
	align-items: center;
	display: flex;
	color: #d44;
	font-size: 1.5rem;
	font-weight: 700;
`;

const rotate = keyframes`
	0% {
		transform: rotate(0deg);
	}

	100% {
		transform: rotate(360deg);
	}
`;

export const Loading = styled.div`
	animation: ${rotate} 0.5s linear infinite;
	border: 0.5rem solid ${({ theme }) => theme.colors.secondary};
	border-left-color: transparent;
	border-radius: 50%;
	border-top-color: transparent;
	height: 4rem;
	margin: auto;
	width: 4rem;
`;
