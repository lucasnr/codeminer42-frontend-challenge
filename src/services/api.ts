import axios from 'axios';

import { IVoucher } from '~/store/modules/cart/types';

const api = axios.create({
	baseURL: 'https://shielded-wildwood-82973.herokuapp.com',
});

export interface IApiProduct {
	id: number;
	name: string;
	price: number;
	available: number;
}

export interface FetchProductsServerResponse {
	products: IApiProduct[];
}

export interface FetchVouchersServerResponse {
	vouchers: IVoucher[];
}

export const getProducts = () =>
	api.get<FetchProductsServerResponse>('products.json');

export const getVouchers = () =>
	api.get<FetchVouchersServerResponse>('vouchers.json');
