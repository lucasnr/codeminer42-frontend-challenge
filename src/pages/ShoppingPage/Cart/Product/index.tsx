import React, { useMemo } from 'react';

import { Container, Content, Name, Row, Controls } from './styles';

import getImage from '~/utils/images';

import { IProduct } from '~/store/modules/cart/types';

const Product: React.FC<IProduct> = ({ id, name, quantity, price }) => {
	const image = useMemo(() => getImage(id), [id]);

	return (
		<Container>
			<img alt={name} src={image} />

			<Content>
				<Name>{name}</Name>
				<Row>
					<span>Quantity: {quantity}</span>
					<span>$ {price}</span>
				</Row>
			</Content>
			<Controls>
				<button>+</button>
				<button>-</button>
			</Controls>
		</Container>
	);
};

export default Product;
