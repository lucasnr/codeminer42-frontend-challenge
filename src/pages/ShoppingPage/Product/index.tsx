import React from 'react';

import { Container, Info, Name, Price, Button } from './styles';

const Product: React.FC = () => {
	return (
		<Container>
			<img
				alt="Product cover"
				src="https://pics.me.me/thumb_kinda-want-this-to-happen-everyone-start-posting-oranges-at-68325797.png"
			/>
			<Info>
				<Name>Product Name</Name>
				<Price>
					<span>$ 123,00</span>
					<div className="dot"></div>
					<span>2 left</span>
				</Price>
			</Info>
			<Button>Buy</Button>
		</Container>
	);
};

export default Product;
