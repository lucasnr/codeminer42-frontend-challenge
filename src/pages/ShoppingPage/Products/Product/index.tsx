import React, { useMemo, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { toast } from 'react-toastify';

import { Container, Info, Name, Row, Button } from './styles';

import { ApplicationState } from '~/store';
import { addProduct } from '~/store/modules/cart/actions';

import { IApiProduct } from '~/services/api';

import getImage from '~/utils/images';
import { formatMoney } from '~/utils/format';

const Product: React.FC<IApiProduct> = ({
	id,
	name,
	price,
	available: initialAvailable,
}) => {
	const image = useMemo(() => {
		return getImage(id);
	}, [id]);

	const cartProducts = useSelector(
		(state: ApplicationState) => state.cart.products
	);
	const available = useMemo(() => {
		const productOnCart = cartProducts.find((product) => product.id === id);
		return initialAvailable - (productOnCart?.quantity || 0);
	}, [initialAvailable, cartProducts, id]);

	const dispatch = useDispatch();
	const handleAddToCart = useCallback(() => {
		if (available < 1) {
			toast.error('This product is no longer available');
			return;
		}

		dispatch(addProduct({ id, name, price, available: initialAvailable }));
	}, [available, dispatch, id, name, price, initialAvailable]);

	const priceFormatted = useMemo(() => formatMoney(price), [price]);

	return (
		<Container>
			<img alt={name} src={image} />
			<Info>
				<Name>{name}</Name>
				<Row>
					<span data-testid="price">$ {priceFormatted}</span>
					<div className="dot"></div>
					<span data-testid="available">{available} left</span>
				</Row>
			</Info>
			<Button onClick={handleAddToCart} disabled={available === 0}>
				{available === 0 ? 'Unavailable' : 'Buy'}
			</Button>
		</Container>
	);
};

export default Product;
