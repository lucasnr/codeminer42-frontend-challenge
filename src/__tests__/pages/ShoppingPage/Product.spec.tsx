import React from 'react';
import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import configureStore, { MockStoreEnhanced } from 'redux-mock-store';

import Product from '~/pages/ShoppingPage/Products/Product';
import { Button } from '~/pages/ShoppingPage/Products/Product/styles';

import { mountWithProviders } from '~/utils/tests';
import { IApiProduct } from '~/services/api';
import { formatMoney } from '~/utils/format';

import { Provider } from 'react-redux';
import realStore, { ApplicationState } from '~/store';

const mockStore = configureStore<ApplicationState>([]);
const initialState: ApplicationState = {
	cart: {
		products: [],
		subtotal: 0,
		shipping: 0,
		discount: 0,
		vouchers: [],
		total: 0,
	},
};

describe('<Product>', () => {
	const product: IApiProduct = {
		id: 5,
		name: 'Strawberry',
		price: 15,
		available: 10,
	};

	let store: MockStoreEnhanced<ApplicationState>;
	let wrapper: ReactWrapper;

	beforeEach(async () => {
		store = mockStore(initialState);

		await act(async () => {
			wrapper = mountWithProviders(<Product {...product} />, store);
		});
	});

	it('renders without fail and matches snapshot', async () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('renders correct values', async () => {
		const formattedPrice = formatMoney(product.price);
		expect(wrapper.find('[data-testid="price"]').text()).toBe(
			`$ ${formattedPrice}`
		);
		expect(wrapper.find('[data-testid="available"]').text()).toBe(
			`${product.available} left`
		);
	});

	it('adds product to store', async () => {
		wrapper = mountWithProviders(
			<Provider store={realStore}>
				<Product {...product} />
			</Provider>
		);
		const button = wrapper.find(Button);

		await act(async () => {
			button.simulate('click');
			wrapper.update();
		});

		expect(realStore.getState().cart.products.length).toBe(1);
		expect(wrapper.find('[data-testid="available"]').text()).toBe(
			`${product.available - 1} left`
		);
	});

	it('does not adds product to store if unavailable', async () => {
		wrapper = mountWithProviders(
			<Provider store={realStore}>
				<Product {...product} available={1} />
			</Provider>
		);
		const button = wrapper.find(Button);

		await act(async () => {
			button.simulate('click');
			wrapper.update();
		});

		expect(realStore.getState().cart.products.length).toBe(1);
		expect(realStore.getState().cart.products[0].quantity).toBe(1);
		expect(wrapper.find('[data-testid="available"]').text()).toBe('0 left');
		expect(button.text()).toBe('Unavailable');

		await act(async () => {
			button.simulate('click');
			wrapper.update();
		});

		expect(realStore.getState().cart.products[0].quantity).toBe(1); // still be 1
	});
});
