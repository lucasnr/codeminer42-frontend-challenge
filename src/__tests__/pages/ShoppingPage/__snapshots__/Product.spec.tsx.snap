// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product> renders without fail and matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ThemeProvider
    theme={
      Object {
        "background": Object {
          "primary": "#fdfdfd",
          "secondary": "#f6f6f6",
        },
        "colors": Object {
          "primary": "#266dd3",
          "secondary": "#344055",
        },
        "text": Object {
          "primary": "#333",
          "secondary": "#fff",
        },
        "title": "light",
      }
    }
  >
    <Product
      available={10}
      id={5}
      name="Strawberry"
      price={15}
    >
      <styled.div>
        <div
          className="sc-AxjAm iYuIGw"
        >
          <img
            alt="Strawberry"
            src="https://dumont.sp.gov.br/lib/img/no-image.jpg"
          />
          <styled.div>
            <div
              className="sc-AxirZ yPRtw"
            >
              <styled.h3>
                <h3
                  className="sc-AxiKw eoCXsE"
                >
                  Strawberry
                </h3>
              </styled.h3>
              <styled.h4>
                <h4
                  className="sc-AxhCb ijKSTM"
                >
                  <span
                    data-testid="price"
                  >
                    $ 
                    15.00
                  </span>
                  <div
                    className="dot"
                  />
                  <span
                    data-testid="available"
                  >
                    10
                     left
                  </span>
                </h4>
              </styled.h4>
            </div>
          </styled.div>
          <styled.button
            disabled={false}
            onClick={[Function]}
          >
            <button
              className="sc-AxhUy ngccn"
              disabled={false}
              onClick={[Function]}
            >
              Buy
            </button>
          </styled.button>
        </div>
      </styled.div>
    </Product>
  </ThemeProvider>
</Provider>
`;
